pkg/file/joiner/internal/job.go

```diff
@@ -0,0 +1,220 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package internal
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"math"
+	"os"
+	"sync"
+
+	"github.com/ethersphere/bee/pkg/file"
+	"github.com/ethersphere/bee/pkg/logging"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/swarm"
+)
+
+// SimpleJoinerJob encapsulates a single joiner operation, providing the consumer
+// with blockwise reads of data represented by a content addressed chunk tree.
+//
+// Every chunk has a span length, which is a 64-bit integer in little-endian encoding
+// stored as a prefix in the chunk itself. This represents the length of the data
+// that reference represents.
+//
+// If a chunk's span length is greater than swarm.ChunkSize, the chunk will be treated
+// as an intermediate chunk, meaning the contents of the chunk are handled as references
+// to other chunks which in turn are retrieved.
+//
+// Otherwise it passes the data chunk to the io.Reader and blocks until the consumer reads
+// the chunk.
+//
+// The process is repeated until the readCount reaches the announced spanLength of the chunk.
+type SimpleJoinerJob struct {
+	ctx           context.Context
+	store         storage.Storer
+	spanLength    int64         // the total length of data represented by the root chunk the job was initialized with.
+	readCount     int64         // running count of chunks read by the io.Reader consumer.
+	cursors       [9]int        // per-level read cursor of data.
+	data          [9][]byte     // data of currently loaded chunk.
+	dataC         chan []byte   // channel to pass data chunks to the io.Reader method.
+	doneC         chan struct{} // channel to signal termination of join loop
+	closeDoneOnce sync.Once     // make sure done channel is closed only once
+	err           error         // read by the main thread to capture error state of the job
+	logger        logging.Logger
+}
+
+// NewSimpleJoinerJob creates a new simpleJoinerJob.
+func NewSimpleJoinerJob(ctx context.Context, store storage.Storer, rootChunk swarm.Chunk) *SimpleJoinerJob {
+	spanLength := binary.LittleEndian.Uint64(rootChunk.Data()[:8])
+	levelCount := getLevelsFromLength(int64(spanLength), swarm.SectionSize, swarm.Branches)
+
+	j := &SimpleJoinerJob{
+		ctx:        ctx,
+		store:      store,
+		spanLength: int64(spanLength),
+		dataC:      make(chan []byte),
+		doneC:      make(chan struct{}, 1),
```

Comment:
could you please comment here why it is needed the buffer the channel? not all that obvious at first sight

Reply:
it's to make sure that the final error state is reported on `Read()`

Reply:
happy to look at a different way to do it over a call (or irl@park tomorrow)